test/integration/apiserver/apiserver_test.go

```diff
@@ -2929,6 +2933,298 @@ func TestDedupOwnerReferences(t *testing.T) {
 	}
 }
 
+func TestEmulatedStorageVersion(t *testing.T) {
+	validVap := &admissionregistrationv1.ValidatingAdmissionPolicy{
+		Spec: admissionregistrationv1.ValidatingAdmissionPolicySpec{
+			MatchConstraints: &admissionregistrationv1.MatchResources{
+				ResourceRules: []admissionregistrationv1.NamedRuleWithOperations{
+					{
+						ResourceNames: []string{"foo"},
+						RuleWithOperations: admissionregistrationv1.RuleWithOperations{
+							Operations: []admissionregistrationv1.OperationType{"CREATE"},
+							Rule: admissionregistrationv1.Rule{
+								APIGroups:   []string{"*"},
+								APIVersions: []string{"*"},
+								Resources:   []string{"*"},
+							},
+						},
+					},
+				},
+			},
+			Validations: []admissionregistrationv1.Validation{
+				{
+					Expression: "true",
+					Message:    "always valid",
+				},
+			},
+		},
+	}
+
+	type testCase struct {
+		name                   string
+		emulatedVersion        string
+		gvr                    schema.GroupVersionResource
+		object                 runtime.Object
+		expectedStorageVersion schema.GroupVersion
+	}
+	cases := []testCase{
+		{
+			name:            "vap first ga release",
+			emulatedVersion: "1.30",
+			gvr: schema.GroupVersionResource{
+				Group:    "admissionregistration.k8s.io",
+				Version:  "v1",
+				Resource: "validatingadmissionpolicies",
+			},
+			object: validVap,
+			expectedStorageVersion: schema.GroupVersion{
+				Group:   "admissionregistration.k8s.io",
+				Version: "v1beta1",
+			},
+		},
+		{
+			name:            "vap after ga release",
+			emulatedVersion: "1.31",
+			gvr: schema.GroupVersionResource{
+				Group:    "admissionregistration.k8s.io",
+				Version:  "v1beta1",
+				Resource: "validatingadmissionpolicies",
+			},
+			object: validVap,
+			expectedStorageVersion: schema.GroupVersion{
+				Group:   "admissionregistration.k8s.io",
+				Version: "v1",
+			},
+		},
+	}
```

Comment:
Clever, but will it keep working in 1.37?  I bet the featuregate look up will struggle handling N-6.

Reply:
this test does not set feature gate. It just depends on the resource definition available in the binary. We keep plenty of resources for a long time after it is annotated as removed. As long as we have this test, it would prevent people from actually removing the file. So it should be ok?

Reply:
that sounds weird. Which part of our flow is allowing a server to startup without lookup featuregates?  And if featuregates are being looked up, why aren't they honor the version?  Seems like integration test authors would expect the emulatedVersion to be honored for feature gates?

Reply:
the resource availability is independent of feature gates (unless they are gated by feature gate somehow?) emulation version not only controls the feature gate, but also the resource availability and storage version. This test is testing the latter. As long as `validatingadmissionpolicies` are still added to the scheme, I think this should work because the emulation version and binary version in the test is fixed.